gempy/core/tensor/tensorflow_graph.py

```diff
@@ -0,0 +1,462 @@
+            self.ref_layer_points[:, 1]),
+            self.gi_reescale ** 2 * (
+            self.rest_layer_points[:, 0] * self.rest_layer_points[:, 2] - self.ref_layer_points[:, 0] *
+            self.ref_layer_points[:, 2]),
+            self.gi_reescale ** 2 * (
+            self.rest_layer_points[:, 1] * self.rest_layer_points[:, 2] - self.ref_layer_points[:, 1] *
+            self.ref_layer_points[:, 2])], 1)
+        
+
+        return U_G, U_I
+
+    def faults_matrix(self, f_ref=None, f_res=None):
+        length_of_CG, _, _, length_of_faults = self.matrices_shapes()[
+            :4]
+
+        # self.fault_drift_at_surface_points_rest = self.fault_matrix
+        # self.fault_drift_at_surface_points_ref = self.fault_matrix
+
+        F_I = (self.fault_drift_at_surface_points_ref -
+               self.fault_drift_at_surface_points_rest) + 0.0001
+
+        F_G = tf.zeros((length_of_faults, length_of_CG),
+                       dtype=self.dtype) + 0.0001
+
+        return F_I, F_G
+
+    @tf.function
+    def covariance_matrix(self):
+
+        length_of_CG, length_of_CGI, length_of_U_I, length_of_faults, length_of_C = self.matrices_shapes()
+
+        C_G = self.cov_gradients()
+        C_I = self.cov_surface_points()
+        C_GI = self.cov_interface_gradients()
+        U_G, U_I = self.universal_matrix()
+        U_G = U_G[:length_of_CG,:3]
+        U_I = U_I[:length_of_CGI,:3]
+        F_I, F_G = self.faults_matrix()
+
+        A = tf.concat([tf.concat([C_G, tf.transpose(C_GI)], -1),
+                       tf.concat([C_GI, C_I], -1)], 0)
+
+        B = tf.concat([U_G, U_I], 0)
+
+        AB = tf.concat([A, B], -1)
+
+        B_T = tf.transpose(B)
+
+        C = tf.pad(B_T, [[0, 0], [0, U_I.shape[1]]])
+
+        C_matrix = tf.concat([AB, C], 0)
+
+
+        C_matrix = tf.where(tf.logical_and(tf.abs(C_matrix) > 0, tf.abs(
+            C_matrix) < 1e-9), tf.constant(0, dtype=self.dtype), y=C_matrix)
+
+        return C_matrix
+
+    def deg2rad(self, degree_matrix):
+        return degree_matrix*tf.constant(0.0174533, dtype=self.dtype)
+
+    @tf.function
+    def b_vector(self, dip_angles_=None, azimuth_=None, polarity_=None):
+
+        length_of_C = self.matrices_shapes()[-1]
+        if dip_angles_ is None:
+            dip_angles_ = self.dip_angles
+        if azimuth_ is None:
+            azimuth_ = self.azimuth
+        if polarity_ is None:
+            polarity_ = self.polarity
+
+        G_x = tf.sin(self.deg2rad(dip_angles_)) * \
+            tf.sin(self.deg2rad(azimuth_)) * polarity_
+        G_y = tf.sin(self.deg2rad(dip_angles_)) * \
+            tf.cos(self.deg2rad(azimuth_)) * polarity_
+        G_z = tf.cos(self.deg2rad(dip_angles_)) * polarity_
+
+        G = tf.concat([G_x, G_y, G_z], -1)
+
+        G = tf.expand_dims(G, axis=1)
+        b_vector = tf.pad(G, [[0, length_of_C - G.shape[0]], [0, 0]])
+
+        return b_vector
+
+    @tf.function
+    def solve_kriging(self, b=None):
```

Comment:
Still got numerical issues for 2nd order drifting. so I switched to 1st order drifting. Compared with Theano results. Similar but not exact due to numerical issues.